Add core WebAssembly tests to WPT

past opened the pull request:
The [Interop 2024 WebAssembly Testing Investigation](https://github.com/web-platform-tests/interop-2024-wasm) intends to get WebAssembly core tests running on WPT.fyi (alongside the existing [JS API](https://wpt.fyi/results/wasm/jsapi?label=experimental&label=master&aligned) and [Web API](https://wpt.fyi/results/wasm/webapi?label=experimental&label=master&aligned) tests).

This PR adds a GitHub Actions workflow that pulls the latest tests from the [WebAssembly/spec](https://github.com/WebAssembly/spec) repo on a regular schedule (or when run manually) and creates a PR with any new changes. 

This mimics the existing process for updating [WebIDL tests](https://github.com/web-platform-tests/wpt/blob/master/.github/workflows/interfaces.yml), which is low risk and low friction. Creating a PR that needs to be reviewed removes the risk of the workflow messing up the wpt repo at the cost of needing manual review of the changes (which is mostly rubber-stamping as in the WebIDL case). I also considered a push model from the spec repo, which could reduce CPU costs and achieve lower latency to propagate changes, but with a substantially more complex solution.

My WPT repo fork has some successful runs that you can look at:

- The [merged PRs](https://github.com/past/wpt/pulls?q=is%3Apr+is%3Aclosed+is%3Amerged) that show up in the repo at [wasm/core](https://github.com/past/wpt/tree/master/wasm/core)
- A [testing PR](https://github.com/past/wpt/pull/7) to verify that the action handles additions, removals and modifications.
- A [test run](https://github.com/past/wpt/actions/runs/11878794719) to ensure the scheduled action succeeds even when no changes exist in WebAssembly/spec 

This should match the MVP that the investigation is pursuing and could subsequently be expanded to cover tentative tests from [WebAssembly/testsuite](https://github.com/WebAssembly/testsuite) or elsewhere. This should let the investigation get to a 67% score in the Interop dashboard.

/CC @jgraham, @gsnedders

past committed:
Regularly add core WebAssembly tests to WPT

dschuff commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+          GIT_AUTHOR_NAME: "wpt-pr-bot"
+          GIT_AUTHOR_EMAIL: "wpt-pr-bot@users.noreply.github.com"
+          BRANCH_PREFIX: "wasm-update"
+          COMMIT_TITLE: "Update WASM tests"
```
```suggestion
          COMMIT_TITLE: "Update Wasm tests"
```

past replied:
Fixed, thanks!

dschuff commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+          gh pr create --title "$COMMIT_TITLE" --body "$PR_BODY" --reviewer past
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          COMMIT_TITLE: "Update WASM tests"
```
```suggestion
          COMMIT_TITLE: "Update Wasm tests"
```

past replied:
Fixed, thanks!

dschuff reviewed (commented):
spelling nits.

past committed:
Fix Wasm spelling

past commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+name: Update WASM tests
```
Fixed, thanks!

past commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+        uses: actions/checkout@v4
+        with:
+          path: main
+      - name: Checkout WASM repo
```
Fixed, thanks!

past commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+        run: cd wasm-spec/interpreter && opam exec make
+      - name: Convert WAST tests to WPT
+        run: wasm-spec/test/build.py --dont-recompile --html wasm-spec/out/
+      - name: Copy WASM tests to WPT
```
Fixed, thanks!

dschuff commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+          rm -rf main/wasm/core
+          cp -r wasm-spec/out/ main/wasm/core/
+          find main/wasm/core/ -type f -name '*.html' -exec sed -i 's/\.\/js\/harness\/testharness/\/resources\/testharness/' {} \;
+      - name: Commit changes
```
If you use the `peter-evans/create-pull-request` action as the WebIDL workflow does, you don't need to manually commit the changes or do anything with git itself.

dschuff replied:
Another example is what we do in [emscripten](https://github.com/emscripten-core/emscripten/blob/main/.github/workflows/tag-release.yml#L20) to create a PR, request a review from a github team, and set the PR to auto-merge (after approval and passing tests)

past replied:
We do something similar in the [wpt-metadata](https://github.com/web-platform-tests/wpt-metadata/tree/master/.github/workflows) repo, but I'm always wary of using 3rd-party libraries for security reasons. However, I see the author of this action is now working for GitHub, so maybe this is less of a concern.

dschuff replied:
Yeah, that makes sense. I'm fine either way, up to you.

Ms2ger reviewed (approved):
Low-confidence (in my ability to review) approval. Can you add a codeowners file to make sure you're tagged on any accidental manual changes?

jgraham commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+      - name: Checkout WPT repo
+        uses: actions/checkout@v4
+        with:
+          path: main
```
Can we call this `wpt` or even `repos/wpt`? I found `main` confusing when I saw it later in the action and had to scroll back up to check it was actually the wpt root.

past replied:
Good point, done.

jgraham commented on .github/workflows/update-wasm-tests.yml:
```diff
@@ -0,0 +1,65 @@
+        run: |
+          rm -rf main/wasm/core
+          cp -r wasm-spec/out/ main/wasm/core/
+          find main/wasm/core/ -type f -name '*.html' -exec sed -i 's/\.\/js\/harness\/testharness/\/resources\/testharness/' {} \;
```
I mean, OK, but this feels like a footgun. Ideally the build step would do this for us, rather than doing string substitution on the built files. If we must do this then we could at least use `'s|input|output|'` or something to avoid needing all the escaping of slashes.

dschuff replied:
I think we can probably fix this in the build script. Let me take a look.

dschuff replied:
I landed https://github.com/WebAssembly/spec/pull/1859 which should match what you have here. I will make further refinements in followup PRs (e.g. there are some harness files that we probably don't need to copy).

past replied:
Thanks for changing build.py Derek! Removing this step is a nice simplification. Tested it in my own fork and it worked flawlessly.

jgraham reviewed (changes requested):
Generally I guess this is OK, just a few small suggestions.

past commented:
> Low-confidence (in my ability to review) approval. Can you add a codeowners file to make sure you're tagged on any accidental manual changes?

I assume you meant to add myself to [`.github/META.yml`](https://github.com/web-platform-tests/wpt/blob/master/.github/META.yml) so that I can be tagged for review, not to add this workflow and myself as the sole reviewer in [`CODEOWNERS`](https://github.com/web-platform-tests/wpt/blob/master/CODEOWNERS)? We don't have any other workflow blocked on core/admin review and this one doesn't seem more critical than the rest.

past committed:
Address review feedback

- Add myself as a reviewer for .github/
- Remove the path fix up step as it's now done by build.py
- Use  better name for the wpt repo directory

Ms2ger commented:
> > Low-confidence (in my ability to review) approval. Can you add a codeowners file to make sure you're tagged on any accidental manual changes?
> 
> I assume you meant to add myself to [`.github/META.yml`](https://github.com/web-platform-tests/wpt/blob/master/.github/META.yml) so that I can be tagged for review, not to add this workflow and myself as the sole reviewer in [`CODEOWNERS`](https://github.com/web-platform-tests/wpt/blob/master/CODEOWNERS)? We don't have any other workflow blocked on core/admin review and this one doesn't seem more critical than the rest.

No, I actually meant something so you get flagged if someone makes changes under `/wasm/core/`, not `/.github`

past commented:
> > > Low-confidence (in my ability to review) approval. Can you add a codeowners file to make sure you're tagged on any accidental manual changes?
> > 
> > 
> > I assume you meant to add myself to [`.github/META.yml`](https://github.com/web-platform-tests/wpt/blob/master/.github/META.yml) so that I can be tagged for review, not to add this workflow and myself as the sole reviewer in [`CODEOWNERS`](https://github.com/web-platform-tests/wpt/blob/master/CODEOWNERS)? We don't have any other workflow blocked on core/admin review and this one doesn't seem more critical than the rest.
> 
> No, I actually meant something so you get flagged if someone makes changes under `/wasm/core/`, not `/.github`

Oh, that would make sense to add to CODEOWNERS. I'll send a PR.

past commented:
> > No, I actually meant something so you get flagged if someone makes changes under `/wasm/core/`, not `/.github`
> 
> Oh, that would make sense to add to CODEOWNERS. I'll send a PR.

Sent #49981.
